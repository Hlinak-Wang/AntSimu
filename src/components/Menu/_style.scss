.menu {
  display: flex;
  flex-wrap: wrap;
  padding-left: 0;
  list-style: none;
  border: $menu-border-width solid $menu-border-color;
  box-shadow: $menu-box-shadow;
  .menu-item {
    padding: $menu-item-padding-y $menu-item-padding-x;
    cursor: pointer;
    transition: $menu-transition;
    &:hover, &:focus {
      text-decoration: none;
    }

    &.is-disabled {
      color:$menu-item-disabled-color;
      pointer-events: none;
      cursor: default;
    }

    &.is-active, &:hover {
      color: $menu-item-active-color;
    }
  }
}

.menu-vertical {
  flex-direction: column;
  margin-bottom: 0px;
  border-right: $menu-border-width solid $menu-border-color;
  > .menu-item {
    border-left: $menu-item-active-border-width solid transparent; 
    &.is-active, &:hover {
      border-left: $menu-item-active-border-width solid $menu-item-active-color;
    }
  }
}

.menu-vertical .submenu {
  border-left: $menu-item-active-border-width solid transparent; 
  padding: $menu-item-padding-y $menu-item-padding-x;
  cursor: pointer;
  transition: $menu-transition;
  position: relative;
  .submenu-item {
    list-style: none;
    padding: 0;
    white-space: nowrap;
    display: none;
    background: white;
  }
  &.menu-open {
    .submenu-item {
      display: block;
    }
  }
  .menu-item {
    list-style: none;
    &.is-active, &:hover {
      color: $menu-item-active-color;
    }
  }
}

.menu-horizontal {
  flex-direction: row;
  margin-bottom: 30px;
  > .menu-item {
    border-bottom: $menu-item-active-border-width solid transparent;
    &.is-active, &:hover {
      border-bottom: $menu-item-active-border-width solid $menu-item-active-color;
    }
  }
}

.menu-horizontal .submenu {
  border-left: $menu-item-active-border-width solid transparent; 
  padding: $menu-item-padding-y $menu-item-padding-x;
  cursor: pointer;
  transition: $menu-transition;
  position: relative;
  .submenu-item {
    position: absolute;
    list-style: none;
    padding: 0;
    white-space: nowrap;
    top: 50px;
    box-shadow: 0 0 5px black;
    display: none;
    background: white;
  }
  &.menu-open {
    .submenu-item {
      display: block;
    }
  }
  .menu-item {
    list-style: none;
    &.is-active, &:hover {
      color: $menu-item-active-color;
    }

    &.is-active {
      background: lighten($primary, 40%);
    }
  }
}



