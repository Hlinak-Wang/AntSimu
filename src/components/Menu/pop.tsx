import React from 'react';

interface PopProps {
  child: React.ReactNode
}
const Pop:React.FC<PopProps> = (props) => {
  const { child } = props
  return (
    <ul className="Pop-itemGroup">
      {child}
    </ul>
  )
}

export default Pop;