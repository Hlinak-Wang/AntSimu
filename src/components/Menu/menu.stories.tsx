import React from 'react';
import Menu from './menu';
import MenuItem from './menuItem';
import SubMenu from './subMenu';
import { storiesOf } from '@storybook/react';
import { actions } from '@storybook/addon-actions';

storiesOf('Menu 菜单', module)
  .add('水平菜单', () => (
    <>
      <p>顶部菜单</p>
      <Menu mode="horizontal">
        <MenuItem handleKey="1">
          123
        </MenuItem>
        <MenuItem handleKey="2">
          123
        </MenuItem>
        <MenuItem handleKey="3">
          123
        </MenuItem>
        <MenuItem handleKey="4">
          123
        </MenuItem>
        <SubMenu title="sub" handleKey="g1">
          <MenuItem handleKey="5">
            123
          </MenuItem>
          <MenuItem handleKey="6">
            123
          </MenuItem>
          <MenuItem handleKey="7">
            123
          </MenuItem>
          <MenuItem handleKey="8">
            123
          </MenuItem>
          <SubMenu title="sub" handleKey="g2">
            <MenuItem handleKey="9">
              123
            </MenuItem>
            <MenuItem handleKey="16">
              123
            </MenuItem>
            <MenuItem handleKey="17">
              123
            </MenuItem>
            <MenuItem handleKey="18">
              123
            </MenuItem>
          </SubMenu>
        </SubMenu>
      </Menu>
    </>
  ))
  .add('垂直菜单', () => (
    <> 
      <Menu mode="vertical" style={{width: "256px", float: "left"}} inline>
      <MenuItem handleKey="1">
          123
        </MenuItem>
        <MenuItem handleKey="2">
          123
        </MenuItem>
        <MenuItem handleKey="3">
          123
        </MenuItem>
        <MenuItem handleKey="4">
          123
        </MenuItem>
        <SubMenu title="sub" handleKey="g1">
          <MenuItem handleKey="5">
            123
          </MenuItem>
          <MenuItem handleKey="6">
            123
          </MenuItem>
          <MenuItem handleKey="7">
            123
          </MenuItem>
          <MenuItem handleKey="8">
            123
          </MenuItem>
        </SubMenu>
      </Menu>
    </>
  ))