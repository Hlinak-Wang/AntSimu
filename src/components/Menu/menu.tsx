import React, { useState, useEffect, createContext } from 'react';
import classnames from 'classnames';
import { MenuItemProps } from './menuItem';

export type MenuMode = 'horizontal' | 'vertical';
export type SubMenuMode = 'pop' | 'inline';
export type SelectCallBack = (selectIndex: string) => void;

export interface MenuProps {
  mode?: MenuMode;
  subMode?: SubMenuMode;
  defaultIndex?: string[];
  className?: string;
  style?:React.CSSProperties;
  onSelect?:SelectCallBack;
  multiple?: boolean;
}
 
interface IMenuContext {
  selectKey: string[];
  onSelect?: SelectCallBack;
  mode?:MenuMode;
  subMode?: SubMenuMode;
}

export const useSelect = (defaultSelect: string[], multiple: boolean, callback?: SelectCallBack) => {
  const [selected, setSelect] = useState<string[]>([...defaultSelect])
  
  const changeSelect = (index: string) => {
    if (multiple) {
      const activeIndex = selected.findIndex(e => e === index);
      if (activeIndex >= 0) {
        setSelect(v => {
          const a = [...v]
          a.splice(activeIndex, 1)
          return a;
        });
      } else {
        setSelect(v => [...v, index]);
      }
    } else {
      setSelect([index]);
    }
    
    callback && callback(index) 
  }
  return {selected, changeSelect}
}

export const MenuContext = createContext<IMenuContext>({selectKey: []})

const Menu: React.FC<MenuProps> = (props) => {
  const {
    mode,
    subMode,
    defaultIndex,
    className,
    style,
    onSelect,
    multiple,
    children
  } = props;

  const {selected, changeSelect} = useSelect(defaultIndex ? defaultIndex : [], multiple!, onSelect);

  useEffect(() => {
    // 检查children是否为MenuItem或SubMenu
    React.Children.forEach(children, (child) => {
      const childElement = child as React.FunctionComponentElement<MenuItemProps>;
      const displayName = childElement.type.displayName || childElement.type.name;
      if (displayName !== 'MenuItem' && displayName !== 'SubMenu') {
        console.error("Menu has a child which is not MenuItem or SubMenu component")
      }
    })
  }, [children]);

  // children的样式
  const passedContext: IMenuContext = {
    selectKey: selected ? [...selected] : [],
    onSelect: changeSelect,
    mode,
    subMode,
  }

  const classes = classnames('menu', className, {
    [`menu-${mode}`]: mode
  });

  return (
    <ul className={classes} data-testid="test-menu" style={style}>
      <MenuContext.Provider value={passedContext}>
        {children}
      </MenuContext.Provider>
    </ul>
  )
}

Menu.defaultProps = {
  mode: 'vertical',
  subMode: 'pop',
  multiple: false,
}
export default Menu;