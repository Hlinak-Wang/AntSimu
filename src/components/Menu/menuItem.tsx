import React, { useContext, useEffect, ReactNode } from 'react';
import classnames from 'classnames';
import { MenuContext } from './menu';
import Animate from '../Animate/animate';

export interface MenuItemProps {
  handleKey: string;
  disabled?:boolean;
  className?:string;
  style?:React.CSSProperties;
  clickRes?: () => void;
  parent?: () => ReactNode;
}

const MenuItem:React.FC<MenuItemProps> = (props) => {
  const {
    handleKey,
    disabled,
    className,
    style,
    children,
    clickRes
  } = props;
  
  const context = useContext(MenuContext);
  const classes = classnames('menu-item', className, {
    'is-disabled': disabled,
    'is-active': context.selectKey.findIndex(v => v === handleKey) >= 0
  });

  const handleClick = () => {
    if (context.onSelect && !disabled) {
      context.onSelect(handleKey)
    }
    if (clickRes) {
      clickRes()
    }
  };
  
  return (
    <li className={classes} style={style} onClick={handleClick}>
      {children}
    </li>
  )
}

export default MenuItem