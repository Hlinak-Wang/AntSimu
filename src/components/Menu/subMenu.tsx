import React, { useContext, useState, useEffect, useRef, CSSProperties, createRef, ReactNode } from 'react';
import classnames from 'classnames';
import { MenuContext } from './menu'
import { MenuItemProps } from './menuItem';
import PopUp from '../popUp/PopUp';
import Icon from '../Icon/icon';
import Animate from '../Animate/animate';
import { CSSTransition } from 'react-transition-group';

interface SubMenuProps {
  handleKey: string;
  className?: string;
  title:string;
  disabled?:boolean;
  parent?: () => ReactNode;
  clickRes?: () => void;
  children?: React.ReactNode[];
}

const SubMenu:React.FC<SubMenuProps> = ({ handleKey, className, title, disabled, clickRes, children, parent }) => {

  const [menuOpen, setOpen] = useState(false);
  const [childSelect, setStatus] = useState(false);
  const context = useContext(MenuContext);
  const subContainRef = createRef<HTMLLIElement>()

  function loopCheck(children: React.ReactNode[]):boolean {
    
    for (let i = 0; i < children.length; i++) {
      let childElement = children[i] as React.FunctionComponentElement<MenuItemProps>;
      let displayName = childElement.type.displayName || childElement.type.name;
      if (displayName === 'SubMenu') {
        let subChild = children[i] as React.FunctionComponentElement<SubMenuProps>;
        let subChildList = React.Children.toArray(subChild.props.children);
        return loopCheck(subChildList);
      } else {
        if (context.selectKey.findIndex(value => value === childElement.props.handleKey) >= 0) {
          return true;
        }
      }
    }
    return false;
  }

  useEffect(() => {
    setStatus(false)
    React.Children.forEach(children, child => {
      const childElement = child as React.FunctionComponentElement<MenuItemProps>;
      const displayName = childElement.type.displayName || childElement.type.name;
      if (displayName === 'SubMenu') {
        const subChild = child as React.FunctionComponentElement<SubMenuProps>;
        const subChildList = React.Children.toArray(subChild.props.children);
        if (loopCheck(subChildList)) {
          setStatus(true);
        }  
      } else {
        if (context.selectKey.findIndex(value => value === childElement.props.handleKey) >= 0) {
          setStatus(true)
        }
      }
    })
  }, [context.selectKey])

  const classes = classnames('submenu', "submenu-title", className, {
    'is-disabled': disabled,
    'menu-open': menuOpen,
    'is-active': childSelect
  })

  // 设置窗口关闭时间间隔
  let timer: any;
  const handleMouse = (toggle: boolean, timeOut: number) => {
    clearTimeout(timer);
    timer = setTimeout(() => {
      setOpen(toggle);
    }, timeOut)
  }

  // const handleClick = (e:React.MouseEvent) => {
  //   e.preventDefault();
  //   setOpen(!menuOpen);
  // }

  // inline 属性
  // const clickEvent = context.inline
  //   ? {onClick: handleClick} 
  //   : {};
  
  const mouseEvent =  {
    onMouseEnter: () => {handleMouse(true, 50)},
    onMouseLeave: () => {handleMouse(false, 50)}
  };

  // 渲染子组件
  const renderChild = () => {
    
    const childrenComponent = React.Children.map(children, (child) => {
      // 检查子组件是否为MenuItem或SubMenu
      // 若为规定的子组件且渲染模式为inline，为其传入handleMouse函数
      const childElement = child as React.FunctionComponentElement<MenuItemProps>;
      const displayName = childElement.type.displayName || childElement.type.name;
      if (displayName === 'MenuItem' || displayName === 'SubMenu') {
        return React.cloneElement(childElement, {
            // 给子级传入handleMouse 函数， 当点击事件在子级触发时执行，并且handleMouse触发以umount父级，并且如果有父级的父级的话在50ms的延迟下再unmount父级的父级
            clickRes: () => {handleMouse(false, 50); setTimeout(() => {clickRes && clickRes()}, 50)},
            parent: () => subContainRef.current
          })
      } else {
        console.error("Menu has a child which is not MenuItem component")
      }
    })

    if (context.subMode == 'pop') {
      return (
        <PopUp 
          getTriggerNode={() => subContainRef.current} 
          popDirec={context.mode == 'horizontal' && !parent ? 'bottom':'right'}
          popUpStyle={{background: "white", boxShadow: "0 0 3px 1px #666", minWidth:"80px"}}
          arrowSize={8}
        >
          <ul
            className="subMenu-payload"
            //className={`subItem-container ${context.inline? 'inline': "outline"}`} 
          >
            {childrenComponent}
          </ul>
        </PopUp>
      )
    }
  }
  
  return (
    <li className={classes} {...mouseEvent} ref={subContainRef} key={handleKey}>
        { title }
        {/* <div className={`arrow-${context.inline || context.mode === 'horizontal' ? "up-down": "left-right"} ${menuOpen ? 'open' : "" }`}/> */}
      { menuOpen && renderChild() }
    </li>
  )
}

export default SubMenu;
