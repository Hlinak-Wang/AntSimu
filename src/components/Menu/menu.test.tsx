import React from 'react';
import { render, RenderResult, fireEvent, cleanup } from '@testing-library/react';
import { renderHook, act } from '@testing-library/react-hooks'
import Menu, { MenuProps, useSelect, SelectCallBack } from './menu';
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import MenuItem from './menuItem';
import subMenu from './subMenu'

const {shallow, mount}=Enzyme
Enzyme.configure({ adapter: new Adapter() })

let testProps: MenuProps;

const testVerProps: MenuProps = {
  defaultIndex: ['0'],
  mode: 'vertical'
}

const generateMenu = (props: MenuProps) => {
  return (
    <Menu {...props}>
      <MenuItem handleKey="0">
        active
      </MenuItem>
      <MenuItem disabled handleKey="1">
        disabled
      </MenuItem>
      <MenuItem handleKey="2">
        not active
      </MenuItem>
    </Menu>
  )
}

//let wrapp: RenderResult, menuElement: HTMLElement, activeElement: HTMLElement, disabledElement: HTMLElement;
let wrapp:Enzyme.ReactWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>
let menuElement: Enzyme.ReactWrapper<Enzyme.HTMLAttributes, any, React.Component<{}, {}, any>>
let activeElement:Enzyme.ReactWrapper<Enzyme.HTMLAttributes, any, React.Component<{}, {}, any>>
let disabledElement:Enzyme.ReactWrapper<Enzyme.HTMLAttributes, any, React.Component<{}, {}, any>>

describe('test Menu and MenuItem', () => {

  beforeEach(() => {
    cleanup()
    testProps = {
      defaultIndex: ['0'],
      onSelect: jest.fn(),
      className: 'test'
    }
    wrapp = mount(generateMenu(testProps))
    menuElement = wrapp.find('.menu')
    activeElement = menuElement.find({handleKey: "0"})
    disabledElement = menuElement.find({handleKey: "1"})
  })
  it('should render correct Menu and MenuItem base on default props', () => {

    expect(menuElement.hasClass(/(?=.*menu)(?=.*menu-vertical).*/)).toBeTruthy()
    expect(menuElement.find('li').length).toEqual(3);
    expect(activeElement.find("li").hasClass(/(?=.*menu-item)(?=.*is-active).*/)).toBeTruthy()
    expect(disabledElement.find('li').hasClass(/(?=.*menu-item)(?=.*is-disabled).*/)).toBeTruthy()
  })
  it('click item should response to the click event', () => {
    const thirdItem = menuElement.find({handleKey: '2'})

    thirdItem.simulate('click')
    expect(testProps.onSelect).toHaveBeenCalledWith('2')

    disabledElement.simulate('click')
    expect(testProps.onSelect).not.toHaveBeenCalledWith('1')
  })
  it('testing useSelect hook in menu for non-mutiple and callback', () => {
    
    const { result } = renderHook(() => useSelect([], false))

    expect(result.current.selected).toEqual([])
    expect(typeof result.current.changeSelect).toBe('function')

    act(() => {
      result.current.changeSelect("1")
    })
    expect(result.current.selected).toEqual(["1"])

    act(() => {
      result.current.changeSelect("2")
    })
    expect(result.current.selected).toEqual(["2"])
  })
  it('testing useSelect hook in menu for multiple and no-callback', () => {
    const { result } = renderHook(() => useSelect([], true))

    expect(result.current.selected).toEqual([])
    expect(typeof result.current.changeSelect).toBe('function')

    act(() => {
      result.current.changeSelect("1")
    })
    expect(result.current.selected).toEqual(["1"])

    act(() => {
      result.current.changeSelect("2")
    })
    expect(result.current.selected).toEqual(["1","2"])
  })
  it('testing useSelect hook in menu for call-back', () => {
    let count = 0;
    const callBack:SelectCallBack = (index: string) => {
      count++;
    }
    const {result} = renderHook(() => useSelect([], true, callBack))
    expect(result.current.selected).toEqual([])
    expect(typeof result.current.changeSelect).toBe('function')

    act(() => {
      result.current.changeSelect("1")
    })
    expect(result.current.selected).toEqual(["1"])
    expect(count).toEqual(1)

    act(() => {
      result.current.changeSelect("2")
    })
    expect(result.current.selected).toEqual(["1","2"])
    expect(count).toEqual(2)
  })
  it('render the correct mode', () => {
    wrapp = mount(generateMenu(testVerProps));
    expect(wrapp).toMatchSnapshot();
  })
}) 