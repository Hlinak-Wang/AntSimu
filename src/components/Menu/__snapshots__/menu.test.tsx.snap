// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Menu and MenuItem render the correct mode 1`] = `ReactWrapper {}`;
